Mathlib/Analysis/SpecialFunctions/Elliptic/Weierstrass.lean

```diff
@@ -0,0 +1,277 @@
+def PeriodPair.latticeEquivProd : L.lattice ≃ₗ[ℤ] ℤ × ℤ :=
+  L.latticeBasis.repr ≪≫ₗ Finsupp.linearEquivFunOnFinite _ _ _ ≪≫ₗ .finTwoArrow ℤ ℤ
+
+lemma PeriodPair.latticeEquiv_symm_apply (x : ℤ × ℤ) :
+    (L.latticeEquivProd.symm x).1 = x.1 * L.ω₁ + x.2 * L.ω₂ := by
+  simp [latticeEquivProd, Finsupp.linearCombination]
+  rfl
+
+open Topology Filter in
+lemma PeriodPair.hasSumLocallyUniformly_aux (f : L.lattice → ℂ → ℂ)
+    (u : ℝ → L.lattice → ℝ) (hu : ∀ r > 0, Summable (u r))
+    (hf : ∀ r > 0, ∀ᶠ R in atTop, ∀ x, ‖x‖ < r → ∀ l : L.lattice, ‖l.1‖ = R → ‖f l x‖ ≤ u r l) :
+    HasSumLocallyUniformlyOn f (∑' j, f j ·) .univ := by
+  rw [hasSumLocallyUniformlyOn_iff_tendstoLocallyUniformlyOn, tendstoLocallyUniformlyOn_univ,
+    tendstoLocallyUniformly_iff_filter]
+  intro x
+  obtain ⟨r, hr, hr'⟩ : ∃ r, 0 < r ∧ 𝓝 x ≤ 𝓟 (Metric.ball 0 r) :=
+    ⟨‖x‖ + 1, by positivity, Filter.le_principal_iff.mpr (Metric.isOpen_ball.mem_nhds (by simp))⟩
+  refine .mono_right ?_ hr'
+  rw [← tendstoUniformlyOn_iff_tendstoUniformlyOnFilter]
+  refine tendstoUniformlyOn_tsum_of_cofinite_eventually (hu r hr) ?_
+  obtain ⟨R, hR⟩ := eventually_atTop.mp (hf r hr)
+  refine (isCompact_iff_finite.mp (isCompact_closedBall (0 : L.lattice) R)).subset ?_
+  intros l hl
+  simp only [Metric.mem_ball, dist_zero_right, Set.mem_compl_iff, Set.mem_setOf_eq, not_forall,
+    not_le, Metric.mem_closedBall, AddSubgroupClass.coe_norm] at hl ⊢
+  obtain ⟨s, hs, hs'⟩ := hl
+  contrapose! hs'
+  exact hR _ hs'.le _ hs _ rfl
+
+lemma PeriodPair.℘_bound (r : ℝ) (hr : r > 0) (s : ℂ) (hs : ‖s‖ < r) (l : ℂ) (h : 2 * r ≤ ‖l‖) :
+    ‖1 / (s - l) ^ 2 - 1 / l ^ 2‖ ≤ 10 * r * ‖l‖ ^ (-3 : ℝ) := by
+  have : s ≠ ↑l := by rintro rfl; exfalso; linarith
+  have : 0 < ‖l‖ := by
+    suffices l ≠ 0 by simpa
+    rintro rfl
+    simp only [norm_zero] at h
+    linarith
+  calc
+    _ = ‖(↑l ^ 2 - (s - ↑l) ^ 2) / ((s - ↑l) ^ 2 * ↑l ^ 2)‖ := by
+      rw [div_sub_div, one_mul, mul_one]
+      · simpa [sub_eq_zero]
+      · simpa
+    _ = ‖l ^ 2 - (s - l) ^ 2‖ / (‖s - l‖ ^ 2 * ‖l‖ ^ 2) := by simp
+    _ ≤ ‖l ^ 2 - (s - l) ^ 2‖ / ((‖l‖ / 2) ^ 2 * ‖l‖ ^ 2) := by
+      gcongr
+      rw [norm_sub_rev]
+      exact .trans (by linarith) (norm_sub_norm_le l s)
+    _ = ‖s * (2 * l - s)‖ / (‖l‖ ^ 4 / 4) := by
+      congr 1
+      · rw [sq_sub_sq]; simp [← sub_add, two_mul, sub_add_eq_add_sub]
+      · ring
+    _ = (‖s‖ * ‖2 * l - s‖) / (‖l‖ ^ 4 / 4) := by simp
+    _ = (4 * ‖s‖ * ‖2 * l - s‖) / ‖l‖ ^ 4 := by field
+    _ ≤ (4 * r * (2.5 * ‖l‖)) / ‖l‖ ^ 4 := by
+      gcongr (4 * ?_ * ?_) / ‖l‖ ^ 4
+      refine (norm_sub_le _ _).trans ?_
+      simp only [Complex.norm_mul, Complex.norm_ofNat]
+      linarith
+    _ = 10 * r / ‖l‖ ^ 3 := by field
+    _ = _ := by norm_cast
+
+section ℘Except
+
+/-- The Weierstrass ℘ function with the `l₀`-term missing. -/
+def PeriodPair.℘Except (l₀ : ℂ) (z : ℂ) : ℂ :=
+  ∑' l : L.lattice, if l.1 = l₀ then 0 else (1 / (z - l) ^ 2 - 1 / l ^ 2)
```

Comment:
Is it worth adding this definition to the Main definitions section? I'm also interested in why it's a useful thing to write down, since my first guess (it's helpful to prove properties about the main function) doesn't appear true from this PR; but maybe it'll show up later?